lib/active_merchant/billing/gateways/worldpay.rb

```diff
@@ -248,10 +248,29 @@ def add_payment_method(xml, amount, payment_method, options)
               xml.tag! 'session', 'shopperIPAddress' => options[:ip] if options[:ip]
               xml.tag! 'session', 'id' => options[:session_id] if options[:session_id]
             end
+            add_stored_credential_options(xml, options)
           end
         end
       end
 
+      def add_stored_credential_options(xml, options={})
+        if merchant_initiated?(xml, options)
+          xml.tag! 'storedCredentials', 'usage' => options[:stored_credential_usage], 'merchantInitiatedReason' => options[:stored_credential_initiated_reason] do
+            xml.tag! 'schemeTransactionIdentifier', options[:stored_credential_transaction_id]
+          end
+        elsif customer_initiated?(xml, options)
+          xml.tag! 'storedCredentials', 'usage' => options[:stored_credential_usage]
+        end
+      end
+
+      def merchant_initiated?(xml, options={})
+        options[:stored_credential_usage] && options[:stored_credential_initiated_reason] && options[:stored_credential_transaction_id]
+      end
+
+      def customer_initiated?(xml, options={})
+        options[:stored_credential_usage]
```

Comment:
Okay, this is subtle, but: given that `merchant_initiated?` and `customer_initiated?` both check for `options[:stored_credential_usage]`, it's possible for `customer_initiated?` to bogusly return `true` (or at least a truthy value). You won't hit that behavior right now, because you're testing for `merchant_iniated?` first, but I can _easily_ see how a superficial refactoring could result in unexpected behavior. So what I'd instead suggest is tweaking this so that the order `customer_initiated?` and `merchant_iniated?` are called in doesn't matter. Something like this:

```suggestion
        !merchant_initiated?(xml, options) && options[:stored_credential_usage]
```

Reply:
In fact I'm not entirely sure we even need/want these methods, rather than just passing through what is present. Currently since they aren't validated in the sandbox, I don't know if partial info would fail or how, but since we're basically leaving all this behavior up to the merchant, I don't think there's much harm in it.